iroh-net/src/magicsock/peer_map/endpoint.rs

```diff
@@ -437,22 +453,35 @@ impl Endpoint {
                 None
             })
             .collect();
-        let ping_needed = !pings.is_empty();
-        let have_endpoints = !self.direct_addr_state.is_empty();
+        let have_ping_msgs = !pings.is_empty();
+        // TODO: Probably don't need this?
+        let have_alive_endpoints = self.direct_addr_state.values().any(|e| e.is_alive());
 
-        if !ping_needed {
+        if !have_ping_msgs {
             trace!("no ping needed");
         }
 
         for ep in pings.into_iter() {
             if let Some(msg) =
                 self.start_ping(SendAddr::Udp(ep.into()), DiscoPingPurpose::Discovery)
             {
-                msgs.push(PingAction::SendPing(msg));
+                ping_msgs.push(PingAction::SendPing(msg));
             }
         }
 
-        if send_call_me_maybe && (ping_needed || !have_endpoints) {
+        // If we were asked for a call-me-maybe we should send it if we don't have a best
+        // addr.  We may not be sending any pings, because we just received a call-me-maybe
+        // and already send pings just now, which remains valid for 5 seconds.
+        //
+        // TODO: double check if there really is any point in still checking beyond
+        // send_call_me_maybe, it posssibly was already checked at the call site that we
+        // don't have a best addr.
+        //
+        // TODO: this could be problematic if we can not punch: we keep sending out
```

Comment:
we should check when we sent the last call-me-maybe probably, and use that as a hint if we need to send one

Reply:
Yep, I think that's the approach i want to take.  see the top-level comment on the pr